### DISCUSSION TITLE
mon.h:43:10: fatal error: 'cstdlib' file not found

### DISCUSSION THREAD

**Discussion by jsevic** [idaholab/moose#25567] [Q&A Modules: Phase field]
I have made a new fresh app called xenon and it currently runs my PF code.

I have added the ExampleConvection .h and .c files and modified the registration to match my app name.

Starting the compiler produces the following error message.

Any suggestions?

mon.h:43:10: fatal error: 'cstdlib' file not found
<img width="836" alt="Screenshot 2023-09-24 at 1 42 21 PM" src="https://github.com/idaholab/moose/assets/4711942/7517fef2-1548-4c13-8d01-5d99d0a8f4b1">

**Comment by GiudGiud** [comment]
Hello

can you run the diagnostics script in moose/scripts?

Guillaume

**Comment by jsevic** [comment]
I tried without and with the .sh extension. Also I am in a zsh shell at the moment.

This is what was returned.

(moose) johnsevic@MBP-16 scripts % diagnostics.sh
zsh: command not found: diagnostics.sh
(moose) johnsevic@MBP-16 scripts % diagnostics   
zsh: command not found: diagnostics

**Reply by GiudGiud** [reply]
`./diagnostics.sh` in that folder

**Comment by jsevic** [comment]
OK thanks here is what the command produced...
```
(moose) johnsevic@MBP-16 scripts % ./diagnostics.sh
Sun Sep 24 17:48:29 PDT 2023

System Arch: Darwin Kernel Version 22.5.0: Mon Apr 24 20:53:19 PDT 2023; root:xnu-8796.121.2~5/RELEASE_ARM64_T6020

MOOSE Package Version: Custom Build

CPU Count: 12

Memory Free: 4845.05 MB

Variable `which $CC` check:
/Users/johnsevic/mambaforge3/envs/moose/bin/mpicc

$CC --version:
clang version 12.0.1
Target: arm64-apple-darwin20.0.0
Thread model: posix
InstalledDir: /Users/johnsevic/mambaforge3/envs/moose/bin

MPICC:
which mpicc:
	/Users/johnsevic/mambaforge3/envs/moose/bin/mpicc
mpicc -show:
	arm64-apple-darwin20.0.0-clang -I/Users/johnsevic/mambaforge3/envs/moose/include -I/Users/johnsevic/mambaforge3/envs/moose/include -L/Users/johnsevic/mambaforge3/envs/moose/lib -Wl,-rpath,/Users/johnsevic/mambaforge3/envs/moose/lib -Wl,-commons,use_dylibs -I/Users/johnsevic/mambaforge3/envs/moose/include -L/Users/johnsevic/mambaforge3/envs/moose/lib -lmpi -lpmpi

COMPILER arm64-apple-darwin20.0.0-clang:
clang version 12.0.1
Target: arm64-apple-darwin20.0.0
Thread model: posix
InstalledDir: /Users/johnsevic/mambaforge3/envs/moose/bin

Python:
	/Users/johnsevic/mambaforge3/envs/moose/bin/python
	Python 3.10.12

MODULES NOT AVAILABLE

PETSc configure:
static const char *petscconfigureoptions = "--with-64-bit-indices --with-cxx-dialect=C++11 --with-debugging=no --with-fortran-bindings=0 --with-mpi=1 --with-openmp=1 --with-shared-libraries=1 --with-sowing=0 --download-fblaslapack=1 --download-hypre=1 --download-metis=1 --download-mumps=1 --download-ptscotch=1 --download-parmetis=1 --download-scalapack=1 --download-slepc=1 --download-strumpack=1 --download-superlu_dist=1 --with-hdf5-dir=${PREFIX} --with-make-np=8 --COPTFLAGS=-O3 --CXXOPTFLAGS=-O3 --FOPTFLAGS=-O3 --with-x=0 --with-ssl=0 --with-mpi-dir=/Users/johnsevic/mambaforge3/envs/moose AR=arm64-apple-darwin20.0.0-ar RANLIB=arm64-apple-darwin20.0.0-ranlib CFLAGS="-ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -isystem /Users/johnsevic/mambaforge3/envs/moose/include   -mcpu=apple-a12" CXXFLAGS="-ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -fmessage-length=0 -isystem ${PREFIX}/include  -mcpu=apple-a12" CPPFLAGS="-D_FORTIFY_SOURCE=2 -isystem /Users/johnsevic/mambaforge3/envs/moose/include -mmacosx-version-min=11.3" FFLAGS="-march=armv8.3-a -ftree-vectorize -fPIC -fno-stack-protector -O2 -pipe -isystem /Users/johnsevic/mambaforge3/envs/moose/include   -march=armv8.3-a" FCFLAGS="-march=armv8.3-a -ftree-vectorize -fPIC -fno-stack-protector -O2 -pipe -isystem /Users/johnsevic/mambaforge3/envs/moose/include   -march=armv8.3-a" LDFLAGS="-Wl,-pie -Wl,-headerpad_max_install_names -Wl,-dead_strip_dylibs -Wl,-rpath,/Users/johnsevic/mambaforge3/envs/moose/lib -L/Users/johnsevic/mambaforge3/envs/moose/lib" --prefix=/Users/johnsevic/mambaforge3/envs/moose";

PETSc linked libraries:
/Users/johnsevic/mambaforge3/envs/moose/lib/libpetsc.dylib:
	@rpath/libpetsc.3.019.dylib (compatibility version 3.19.0, current version 3.19.3)
	@rpath/libHYPRE-2.29.0.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
	@rpath/libstrumpack.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libscalapack.2.2.dylib (compatibility version 2.2.0, current version 2.2.1)
	@rpath/libsuperlu_dist.8.dylib (compatibility version 8.0.0, current version 8.1.2)
	@rpath/libparmetis.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libmetis.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libhdf5.200.dylib (compatibility version 203.0.0, current version 203.0.0)
	@rpath/libmpifort.12.dylib (compatibility version 15.0.0, current version 15.2.0)
	@rpath/libmpi.12.dylib (compatibility version 15.0.0, current version 15.2.0)
	@rpath/libgfortran.5.dylib (compatibility version 6.0.0, current version 6.0.0)
	@rpath/libc++.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libomp.dylib (compatibility version 5.0.0, current version 5.0.0)

ENVIRONMENT:
AR=arm64-apple-darwin20.0.0-ar
AS=arm64-apple-darwin20.0.0-as
BUILD=arm64-apple-darwin20.0.0
CC=mpicc
CC_FOR_BUILD=/Users/johnsevic/mambaforge3/envs/moose/bin/arm64-apple-darwin20.0.0-clang
CFLAGS=-ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -isystem /Users/johnsevic/mambaforge3/envs/moose/include
CHECKSYMS=arm64-apple-darwin20.0.0-checksyms
CLANG=arm64-apple-darwin20.0.0-clang
CLANGXX=arm64-apple-darwin20.0.0-clang++
CMAKE_ARGS=-DCMAKE_AR=/Users/johnsevic/mambaforge3/envs/moose/bin/arm64-apple-darwin20.0.0-ar -DCMAKE_CXX_COMPILER_AR=/Users/johnsevic/mambaforge3/envs/moose/bin/arm64-apple-darwin20.0.0-ar -DCMAKE_C_COMPILER_AR=/Users/johnsevic/mambaforge3/envs/moose/bin/arm64-apple-darwin20.0.0-ar -DCMAKE_RANLIB=/Users/johnsevic/mambaforge3/envs/moose/bin/arm64-apple-darwin20.0.0-ranlib -DCMAKE_CXX_COMPILER_RANLIB=/Users/johnsevic/mambaforge3/envs/moose/bin/arm64-apple-darwin20.0.0-ranlib -DCMAKE_C_COMPILER_RANLIB=/Users/johnsevic/mambaforge3/envs/moose/bin/arm64-apple-darwin20.0.0-ranlib -DCMAKE_LINKER=/Users/johnsevic/mambaforge3/envs/moose/bin/arm64-apple-darwin20.0.0-ld -DCMAKE_STRIP=/Users/johnsevic/mambaforge3/envs/moose/bin/arm64-apple-darwin20.0.0-strip -DCMAKE_INSTALL_NAME_TOOL=/Users/johnsevic/mambaforge3/envs/moose/bin/arm64-apple-darwin20.0.0-install_name_tool -DCMAKE_LIBTOOL=/Users/johnsevic/mambaforge3/envs/moose/bin/arm64-apple-darwin20.0.0-libtool -DCMAKE_OSX_DEPLOYMENT_TARGET= -DCMAKE_BUILD_TYPE=Release -DCMAKE_OSX_SYSROOT=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk
CMAKE_PREFIX_PATH=:/Users/johnsevic/mambaforge3/envs/moose
CONDA_BACKUP_HOST=MBP-16.local
CONDA_DEFAULT_ENV=moose
CONDA_EXE=/Users/johnsevic/mambaforge3/bin/conda
CONDA_PREFIX=/Users/johnsevic/mambaforge3/envs/moose
CONDA_PREFIX_1=/Users/johnsevic/mambaforge3
CONDA_PROMPT_MODIFIER=(moose) 
CONDA_PYTHON_EXE=/Users/johnsevic/mambaforge3/bin/python
CONDA_SHLVL=2
CONDA_TOOLCHAIN_BUILD=arm64-apple-darwin20.0.0
CONDA_TOOLCHAIN_HOST=arm64-apple-darwin20.0.0
CPPFLAGS=-D_FORTIFY_SOURCE=2 -isystem /Users/johnsevic/mambaforge3/envs/moose/include
CXX=mpicxx
CXXFLAGS=-ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -fmessage-length=0 -isystem /Users/johnsevic/mambaforge3/envs/moose/include -std=c++17
CXX_FOR_BUILD=/Users/johnsevic/mambaforge3/envs/moose/bin/arm64-apple-darwin20.0.0-clang++
C_INCLUDE_PATH=/Users/johnsevic/mambaforge3/envs/moose/include
DEBUG_CFLAGS=-ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -Og -g -Wall -Wextra -isystem /Users/johnsevic/mambaforge3/envs/moose/include
DEBUG_CXXFLAGS=-ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -fmessage-length=0 -Og -g -Wall -Wextra -isystem /Users/johnsevic/mambaforge3/envs/moose/include
DEBUG_FFLAGS=-march=armv8.3-a -ftree-vectorize -fPIC -fno-stack-protector -O2 -pipe -isystem /Users/johnsevic/mambaforge3/envs/moose/include -march=armv8.3-a -ftree-vectorize -fPIC -fno-stack-protector -O2 -pipe -Og -g -Wall -Wextra -fcheck=all -fbacktrace -fimplicit-none -fvar-tracking-assignments
DEBUG_FORTRANFLAGS=-march=armv8.3-a -ftree-vectorize -fPIC -fno-stack-protector -O2 -pipe -isystem /Users/johnsevic/mambaforge3/envs/moose/include -march=armv8.3-a -ftree-vectorize -fPIC -fno-stack-protector -O2 -pipe -Og -g -Wall -Wextra -fcheck=all -fbacktrace -fimplicit-none -fvar-tracking-assignments
Eigen3_DIR=/Users/johnsevic/mambaforge3/envs/moose/libmesh/include/Eigen
F77=mpif77
F90=mpif90
F95=/Users/johnsevic/mambaforge3/envs/moose/bin/arm64-apple-darwin20.0.0-gfortran
FC=mpif90
FC_FOR_BUILD=/Users/johnsevic/mambaforge3/envs/moose/bin/arm64-apple-darwin20.0.0-gfortran
FFLAGS=-march=armv8.3-a -ftree-vectorize -fPIC -fno-stack-protector -O2 -pipe -isystem /Users/johnsevic/mambaforge3/envs/moose/include
FORTRANFLAGS=-march=armv8.3-a -ftree-vectorize -fPIC -fno-stack-protector -O2 -pipe -isystem /Users/johnsevic/mambaforge3/envs/moose/include
GFORTRAN=/Users/johnsevic/mambaforge3/envs/moose/bin/arm64-apple-darwin20.0.0-gfortran
GSETTINGS_SCHEMA_DIR=/Users/johnsevic/mambaforge3/envs/moose/share/glib-2.0/schemas
GSETTINGS_SCHEMA_DIR_CONDA_BACKUP=
HDF5_DIR=/Users/johnsevic/mambaforge3/envs/moose
HOME=/Users/johnsevic
HOST=arm64-apple-darwin20.0.0
INSTALL_NAME_TOOL=arm64-apple-darwin20.0.0-install_name_tool
LANG=en_US.UTF-8
LD=arm64-apple-darwin20.0.0-ld
LDFLAGS=-Wl,-pie -Wl,-headerpad_max_install_names -Wl,-dead_strip_dylibs -Wl,-rpath,/Users/johnsevic/mambaforge3/envs/moose/lib -L/Users/johnsevic/mambaforge3/envs/moose/lib
LDFLAGS_LD=-pie -headerpad_max_install_names -dead_strip_dylibs -rpath /Users/johnsevic/mambaforge3/envs/moose/lib -L/Users/johnsevic/mambaforge3/envs/moose/lib
LIBMESH_DIR=/Users/johnsevic/mambaforge3/envs/moose/libmesh
LIBTOOL=arm64-apple-darwin20.0.0-libtool
LIPO=arm64-apple-darwin20.0.0-lipo
LOGNAME=johnsevic
MESON_ARGS=--buildtype release
MOOSE_NO_CODESIGN=true
MPIHOME=/Users/johnsevic/mambaforge3/envs/moose
NM=arm64-apple-darwin20.0.0-nm
NMEDIT=arm64-apple-darwin20.0.0-nmedit
OBJC=arm64-apple-darwin20.0.0-clang
OBJC_FOR_BUILD=/Users/johnsevic/mambaforge3/envs/moose/bin/arm64-apple-darwin20.0.0-clang
OTOOL=arm64-apple-darwin20.0.0-otool
PAGESTUFF=arm64-apple-darwin20.0.0-pagestuff
PATH=/Users/johnsevic/mambaforge3/envs/moose/bin:/Users/johnsevic/mambaforge3/condabin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/TeX/texbin:/Library/Apple/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Users/johnsevic/mambaforge3/envs/moose/wasp/bin:/Users/johnsevic/mambaforge3/envs/moose/wasp/bin:/Users/johnsevic/mambaforge3/envs/moose/wasp/bin
PETSC_DIR=/Users/johnsevic/mambaforge3/envs/moose
PROJ_DATA=/Users/johnsevic/mambaforge3/envs/moose/share/proj
PROJ_NETWORK=ON
PWD=/Users/johnsevic/projects/moose/scripts
RANLIB=arm64-apple-darwin20.0.0-ranlib
REDO_PREBINDING=arm64-apple-darwin20.0.0-redo_prebinding
SDKROOT=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk
SEGEDIT=arm64-apple-darwin20.0.0-segedit
SEG_ADDR_TABLE=arm64-apple-darwin20.0.0-seg_addr_table
SEG_HACK=arm64-apple-darwin20.0.0-seg_hack
SHELL=/bin/zsh
SHLVL=2
SIZE=arm64-apple-darwin20.0.0-size
SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.zdavRKgD7E/Listeners
STRINGS=arm64-apple-darwin20.0.0-strings
STRIP=arm64-apple-darwin20.0.0-strip
TERM=xterm-256color
TERM_PROGRAM=Apple_Terminal
TERM_PROGRAM_VERSION=447
TERM_SESSION_ID=62687E71-AE40-403C-B77E-A2FEE494B9A9
TMPDIR=/var/folders/nm/kzlqqypx4bz00c5f3d9nf5d40000gn/T/
USER=johnsevic
VTKINCLUDE_DIR=/Users/johnsevic/mambaforge3/envs/moose/libmesh-vtk/include/vtk-9.2
VTKLIB_DIR=/Users/johnsevic/mambaforge3/envs/moose/libmesh-vtk/lib
WASP_DIR=/Users/johnsevic/mambaforge3/envs/moose/wasp
XML_CATALOG_FILES=file:///Users/johnsevic/mambaforge3/envs/moose/etc/xml/catalog file:///etc/xml/catalog
XPC_FLAGS=0x0
XPC_SERVICE_NAME=0
_=/usr/bin/env
_CE_CONDA=
_CE_M=
_CONDA_PYTHON_SYSCONFIGDATA_NAME=_sysconfigdata_arm64_apple_darwin20_0_0
__CFBundleIdentifier=com.apple.Terminal
ac_cv_func_malloc_0_nonnull=yes
ac_cv_func_realloc_0_nonnull=yes
build_alias=arm64-apple-darwin20.0.0
host_alias=arm64-apple-darwin20.0.0
```

**Reply by GiudGiud** [reply]
@milljm see anything odd?

**Reply by milljm** [reply]
The only odd ball I see is

```pre
SDKROOT=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk
```
being set. Can you try unsetting that, and running make again?

```bash
mamba activate moose
unset SDKROOT
# cd into your xenon directory and then:
make -j 12
```
If that fails, try cleaning out the old build and start anew (see careful note below before proceeding):
```bash
mamba activate moose
unset SDKROOT
# cd into your xenon directory and then:
make clobberall
make -j 12
```
* Be careful with `make clobberall`. Save your work before running. I think it has the potential of deleting uncommitted work.

**Reply by jsevic** [reply]
OK thanks I will try. FYI the app proceeding this one, radon, has been working on phase field the last couple of months and I tried adding to this app last week the ExampleConvection kernel, in preparation for making a new kernel. When I compiled that app I got the same response as this one in xenon. Is there an issue maybe with my terminal shell? Often conda or mamba won't activate MOOSE I notice, when I open a new terminal?

**Reply by GiudGiud** [reply]
>  Often conda or mamba won't activate MOOSE

what happens?
mamba activate moose fails?
Or the shell will start and it s not activated by default?

**Comment by jsevic** [comment]
If I open a new terminal and type mamba activate moose I get a run mamba init and then I run that an reopen a new terminal and the process seems to repeat. It's not clear to me the new terminal retains the new configuration?

I tried the unset SDKROOT and clobberall and I get compile errors in both cases, I can post these if you like?

**Reply by GiudGiud** [reply]
which shell are you using?

**Reply by GiudGiud** [reply]
oh zsh.

Yeah so there's an issue with mamba and zsh. 
You ll need to open your ~/.zshrc file and paste the stuff that mamba added to your .bashprofile

**Comment by jsevic** [comment]
<img width="891" alt="Screenshot 2023-09-25 at 12 29 18 PM" src="https://github.com/idaholab/moose/assets/4711942/26351be1-0684-4efc-ae0d-c713ffa6fe17">

**Reply by milljm** [reply]
this is happening because the 'moose' Conda environment isn't active. And by reading the previous posts, your machine seems to have an issue activating these environments. What you need to do is manually edit your `~/.zshrc` file and add the following to the end of it:

```pre
# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/Users/johnsevic/mambaforge3/bin/conda' 'shell.zsh' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/Users/johnsevic/mambaforge3/etc/profile.d/conda.sh" ]; then
        . "/Users/johnsevic/mambaforge3/etc/profile.d/conda.sh"
    else
        export PATH="/Users/johnsevic/mambaforge3/bin:$PATH"
    fi
fi
unset __conda_setup

if [ -f "/Users/johnsevic/mambaforge3/etc/profile.d/mamba.sh" ]; then
    . "/Users/johnsevic/mambaforge3/etc/profile.d/mamba.sh"
fi
# <<< conda initialize <<<
```

Once you do this, close all your terminals, and start them back up. You should see a `(base)` prefix in your prompt. Once you do, you should then be able to activate the moose environment:

```bash
mamba activate moose
```

And then perform `make` again.

**Reply by milljm** [reply]
Reading the entire thread over again, I am a bit confused how you managed to run into a `cstdlib` issue (while clearly using the `moose` Conda environment by way of those paths being listed), but in later posts, activating the moose Conda environment is problematic. Can you elaborate on how this happened?

**Reply by jsevic** [reply]
OK thanks, so I added the above script to my .zshrc file (in fact it's empty except for this script) and then I killed the terminal app and restarted it. I get base. And mamba activate moose seems to work (at least no errors and I get the command line back). I run this from ~/projects.

From there I go into xenon and run make -j 12 and then I get the following same error.

As to why this happens, this is a good question, it seems like everything works fine for compiling then once I close the terminal I start getting compile errors.

<img width="886" alt="Screenshot 2023-09-25 at 1 42 13 PM" src="https://github.com/idaholab/moose/assets/4711942/540f53c5-67f5-49a1-9f89-a12bf25d98a4">


When I run

**Reply by jsevic** [reply]
Honestly I cannot say what causes this to happen. I have been using MOOSE since 2017, and I get often got this error, and my workaround was to create a new app. That's why I tried to do here. Something tells me there is an issue with Conda and Mamba, and my shell configuration, but I cannot get my hands around it. For example, radon has been working fine since August and last week I added ExampleConvection to it and opened a new shell and this the error I got. So I made xenon, a fresh MOOSE install and new app, and still got the same error. What does the error mean?

I am willing to try anything let me know thx!

John Sevic

**Comment by jsevic** [comment]
Also I was able to compile and run all the tests on moose, as well as compile ex2 in examples. If I next attempt to compile in xenon, I get the cstlib error.

**Comment by jsevic** [comment]
I did a fresh install for a new app named argon. It passes tests. I can run my PF code. When I added the ExampleConvection.h and ExampleConvection.c and compile I get the cstdlib file not found error as before.


<img width="942" alt="Screenshot 2023-09-25 at 6 31 29 PM" src="https://github.com/idaholab/moose/assets/4711942/67f9bb67-a87f-4c49-abee-a536c652fd51">

**Reply by milljm** [reply]
please stop submitting pictures. It does little good in the way of others searching for others who may run into "fatal error: 'cstdlib' file not found" when that text is only contained in pictures.

**Reply by milljm** [reply]
The only thing I can think of, is something wrong with XCode you have installed. What version are you using?

**Comment by jsevic** [comment]
As I said in the last post that you may have missed I can compile. Tests run. PF runs. When I for the the third time make a new app which passes tests, and compile ExampleConvection, the same cstslib error returns. So XCode seems to be working. I saw this on Intel and Apple Si for several years already.

What, exactly, does cstdlib do? It's in libmesh it appears? Is there a workaround to get this file into the location it needs to be in?

Thank you. 

John Sevic

**Reply by milljm** [reply]
libmesh_common.h is expecting to find `cstdlib` globally. So it is odd that it cannot be found.

normally... this falls to an Xcode and/or Command Line Tools issues. You are the only one I've run into who has (had) an SDKROOT environment variable set. Its the only oddity I have seen so far. They are probably related... somehow.

@roystgnr What do you think this could be?

**Reply by milljm** [reply]
While Conda appears to be priority in your path, I do see a `/usr/local/bin` there. Sometimes, binaries/libraries in the `/usr/local` location can interfere. Can you list what is available in the following locations (if they exist):

```bash
ls /usr/local/bin
ls /usr/local/lib
ls /usr/local/include
```

**Reply by roystgnr** [reply]
It's definitely a broken compiler.  This isn't a libMesh header; it's not even `rpc.h` (a POSIX standard with not-quite-standard locations).  If a "C++ compiler" fails to find the C++ standard `cstdlib` then it's [by definition](https://cplusplus.com/reference/cstdlib/) not a C++ compiler after all.

As to *how* it's broken... that I have no idea.  I'm a Linux person, and on any distribution with a package manager you'll have `cstdlib` brought in by mandatory dependencies: e.g. `apt install g++` brings in `g++-12` which brings in `libstdc++-12-dev` which installs `/usr/include/c++/12/cstdlib`.

On my Macbook from `locate cstdlib` I'm seeing versions in  `/Library/Developer/CommandLineTools/usr/include/c++/v1`, as well as in 4 or 5 different conda packages.  Running `clang++ -E -x c++ - -v` with any installed `clang++` shows one of those places hard-coded into the compiler, as it's supposed to be.

@jsevic, what do the `locate cstdlib` and `clang++ -E -x c++ - -v` (assuming that's the compiler you're using) commands say for you?

**Reply by jsevic** [reply]
The locate command is building so I will wait for that tonight. The clang++ command produced the following output...

(moose) johnsevic@MBP-16 argon % clang++ -E -x c++ - -v
clang version 12.0.1
Target: arm64-apple-darwin22.5.0
Thread model: posix
InstalledDir: /Users/johnsevic/mambaforge3/envs/moose/bin
 (in-process)
 "/Users/johnsevic/mambaforge3/envs/moose/bin/clang-12" -cc1 -triple arm64-apple-macosx13.0.0 -Wundef-prefix=TARGET_OS_ -Werror=undef-prefix -Wdeprecated-objc-isa-usage -Werror=deprecated-objc-isa-usage -E -disable-free -disable-llvm-verifier -discard-value-names -main-file-name - -mrelocation-model pic -pic-level 2 -mframe-pointer=non-leaf -fno-rounding-math -munwind-tables -target-sdk-version=13.3 -fcompatibility-qualified-id-block-type-checking -fvisibility-inlines-hidden-static-local-var -target-cpu apple-a12 -target-feature +v8.3a -target-feature +fp-armv8 -target-feature +neon -target-feature +crc -target-feature +crypto -target-feature +fullfp16 -target-feature +ras -target-feature +lse -target-feature +rdm -target-feature +rcpc -target-feature +zcm -target-feature +zcz -target-feature +sha2 -target-feature +aes -target-abi darwinpcs -fallow-half-arguments-and-returns -debugger-tuning=lldb -target-linker-version 530 -v -resource-dir /Users/johnsevic/mambaforge3/envs/moose/lib/clang/12.0.1 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -c-isystem /Users/johnsevic/mambaforge3/envs/moose/include -stdlib=libc++ -internal-isystem /Users/johnsevic/mambaforge3/envs/moose/bin/../include/c++/v1 -internal-isystem /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/local/include -internal-isystem /Users/johnsevic/mambaforge3/envs/moose/lib/clang/12.0.1/include -internal-externc-isystem /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include -fdeprecated-macro -fdebug-compilation-dir /Users/johnsevic/projects/argon -ferror-limit 19 -stack-protector 1 -fblocks -fencode-extended-block-signature -fregister-global-dtors-with-atexit -fgnuc-version=4.2.1 -fcxx-exceptions -fexceptions -fmax-type-align=16 -fcolor-diagnostics -o - -x c++ -
clang -cc1 version 12.0.1 based upon LLVM 12.0.1 default target arm64-apple-darwin22.5.0
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/local/include"
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/Library/Frameworks"
#include "..." search starts here:
#include <...> search starts here:
 /Users/johnsevic/mambaforge3/envs/moose/bin/../include/c++/v1
 /Users/johnsevic/mambaforge3/envs/moose/lib/clang/12.0.1/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks (framework directory)
End of search list.

**Reply by jsevic** [reply]
Here is what I got from these commands...

(moose) johnsevic@MBP-16 ~ % ls /usr/local/bin

dvipdf		gsdj500		pf2afm		ps2pdf12
eps2eps		gslj		pfbtopfa	ps2pdf13
gs		gslp		pphs		ps2pdf14
gs-X11		gsnd		printafm	ps2pdfwr
gs-noX11	lprsetup.sh	ps2ascii	ps2ps
gsbj		pdf2dsc		ps2epsi		ps2ps2
gsdj		pdf2ps		ps2pdf		unix-lpr.sh
(moose) johnsevic@MBP-16 ~ % ls /usr/local/lib

ls: /usr/local/lib: No such file or directory
(moose) johnsevic@MBP-16 ~ % ls /usr/local/include

ls: /usr/local/include: No such file or directory
(moose) johnsevic@MBP-16 ~ %

**Reply by roystgnr** [reply]
`ignoring nonexistent directory "/Applications/Xcode.app/` etc etc is scary.  I'm not sure exactly how our conda clang-12 was configured but I'm guessing some part of the process is expecting those directories and a bunch of header files within them to exist.

**Comment by jsevic** [comment]
What I have done traditionally as a workaround is to make my app just the basic moose install versus creating the preferred new app. Usually moose compiles and then I add PF and heat conduction and it's ok. It's creating a new app that seems to make an issue.  I will try tonight adding PF and heat conduction to moose and see what happens.

What about a clean install? Is there a way to ensure "everything" gets removed and won't interfere with a fresh install?

Thx!

**Comment by jsevic** [comment]
It seems to me make clobberall may be causing unexpected behavior. I went to \examples\ex02_kernel and compiles no problems.

Then I added to the makefile the flags to add phase field and heat conduction, and ran the make file again but it evidently did not register the two modules, so I ran make clobberall and then ran make -j 12 and now I get compile errors even after setting PF and heat conduction flags to no, i.e. returning the make file to its original state.

Any comments or suggestions?

**Reply by roystgnr** [reply]
Can you compile a basic program using `cstdlib` outside MOOSE?  E.g. if you create a `foo.cc` containing:

```
#include <cstdlib>
#include <iostream>
int main(void) {
  std::cout << std::strtod("2",nullptr)*std::strtod("3",nullptr) << std::endl;
  return 0;
}
```

And you just `clang++ -o foo.x foo.cc && ./foo.x`, does it then print "6" with no errors?

**Comment by jsevic** [comment]
I will try tonight.At the moment I can compile fine in Moose, but I have to use the app moose. My other apps  xenon and radon do not compile.Thx. John SevicSent from my iPhoneOn Sep 27, 2023, at 5:46 AM, roystgnr ***@***.***> wrote:﻿
Can you compile a basic program using cstdlib outside MOOSE?  E.g. if you create a foo.cc containing:
#include <cstdlib>
#include <iostream>
int main(void) {
  std::cout << std::strtod("2",nullptr)*std::strtod("3",nullptr) << std::endl;
  return 0;
}

And you just clang++ -o foo.x foo.cc && ./foo.x, does it then print "6" with no errors?
